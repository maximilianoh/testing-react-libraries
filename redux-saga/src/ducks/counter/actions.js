import types from './types';

const actions = {
    increase: () => ({ type: types.INCREMENT}),
    decrease: () => ({ type: types.DECREMENT}),
    reset: () => ({ type: types.RESET}),
    getData: () => ({ type: types.GET_TODOS_REQUESTED}),
  };
  
export default actions;