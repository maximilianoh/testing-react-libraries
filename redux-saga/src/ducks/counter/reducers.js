import types from './types';
const INITIAL_STATE = {
    count: 0,
    history: [],
    data:[],
}
  
function handleChange(state, change) {
  const {count, history} = state;
  return ({
    ...state,
    count: count + change,
    history: [count + change, ...history],
  })
}

function apiChange(state, payload){
  return ({
    ...state,
    data: payload,
  })
}

export default function counter(state = INITIAL_STATE, action) {
    switch(action.type) {
      case types.INCREMENT:
        return handleChange(state, 1);
      case types.DECREMENT:
        return handleChange(state, -1);
      case types.RESET:
        return (INITIAL_STATE);
      case types.GET_TODOS:
        return apiChange(state, action.payload);
      default:
        return state;
    }
}
