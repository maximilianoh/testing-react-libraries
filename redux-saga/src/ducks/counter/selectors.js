const getCount = (state) => state.counterGroup.count;
const getHistory = (state) => state.counterGroup.history;
const getSize = (state) => state.counterGroup.data.length;
const selectors = {
    getCount,
    getHistory,
    getSize
};
export default selectors;