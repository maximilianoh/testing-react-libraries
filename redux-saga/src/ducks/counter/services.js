import defaultAxios from 'axios'


const axios = defaultAxios.create({
  baseURL: 'https://jsonplaceholder.typicode.com/',
  headers: {'Content-Type': 'application/json'}
});

// Get All Todos
export const getAllTodos = async () => {
  try {
    const todos = await axios.get('todos?_limit=5');
    return todos.data;
  } catch(err) {
    return console.error(err);
  }
}
  