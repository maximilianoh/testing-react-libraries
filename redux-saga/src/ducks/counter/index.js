import actions from './actions';
import reducer from './reducers';
import types from './types';
import selectors from './selectors';

export {
  actions,
  reducer,
  selectors,
  types
};