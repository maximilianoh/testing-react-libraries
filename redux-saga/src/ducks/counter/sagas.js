import { call, takeLatest, put } from 'redux-saga/effects';
import {getAllTodos} from './services';
import types from './types';


// Get Todos
function* getTodos() {
    const todos = yield call(getAllTodos);
    yield put({ type: types.GET_TODOS, payload: todos })
}

// Export the saga (todo-saga)
function* mySaga() {
    yield takeLatest(types.GET_TODOS_REQUESTED, getTodos);
  }
export default mySaga;
