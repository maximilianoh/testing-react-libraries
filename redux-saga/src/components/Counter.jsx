import { useDispatch, useSelector } from 'react-redux';
import { counterGroup } from '../ducks/';

function App() {
  const dispatch = useDispatch();
  const counter = useSelector(counterGroup.selectors.getCount);
  return (
    <div> 
        <h1> Counter </h1>
        <h1> {counter} </h1>
        <button onClick={() => dispatch(counterGroup.actions.increase())}> Increase </button>
        <button onClick={() => dispatch(counterGroup.actions.decrease())}> Decrease </button>
        <button onClick={() => dispatch(counterGroup.actions.reset())}> Reset </button>
    </div>
  );
}

export default App;