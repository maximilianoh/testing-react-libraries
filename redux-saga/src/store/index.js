import { createStore, combineReducers, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga'
import * as ServiceGroups from '../ducks';
import rootSaga from '../ducks/counter/sagas';
// Se arma el objeto con todos los reducers { ExampleServiceGroup: reducer, ... }
const reducers = Object.fromEntries(
    Object.entries(ServiceGroups)
      .map(([key, value]) => ([
        key, value.reducer
      ]))
);

// Se arma el array con todos los sagas
/*const sagas = Object.entries(ServiceGroups)
  .reduce((acc, [, value]) => acc.concat(...value.sagas), []);*/
 

/* ------------- Assemble The Reducers ------------- */
const rootReducer = combineReducers(reducers);

// create the saga middleware
const sagaMiddleware = createSagaMiddleware()

let store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

// then run the saga
sagaMiddleware.run(rootSaga)

export default store;