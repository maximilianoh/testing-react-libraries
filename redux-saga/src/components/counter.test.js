import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import Counter from './Counter';
import { Provider } from 'react-redux'
import store from '../store';
import { reducer, types } from '../ducks/counter';

test('Default Snap', () => {
    const component = TestRenderer.create(
        <Provider store={store}>
            <Counter/>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});


test('Default Increase', () => {
    const { rerender, baseElement } = render(
        <Provider store={store}>
            <Counter/>
        </Provider>
    );
    expect(screen.queryByText('0')).toBeInTheDocument()
    const {act} = TestRenderer;
    act(() => {
        fireEvent.click(screen.getByText('Increase'));
    });
    expect(screen.queryByText('0')).toBe(null)
    expect(screen.queryByText('1')).toBeInTheDocument();
    act(() => {
        fireEvent.click(screen.getByText('Increase'));
    });
    expect(screen.queryByText('1')).toBe(null)
    expect(screen.queryByText('2')).toBeInTheDocument();
    act(() => {
        fireEvent.click(screen.getByText('Decrease'));
    });
    expect(screen.queryByText('2')).toBe(null)
    expect(screen.queryByText('1')).toBeInTheDocument();
});

test('Default reduce Initial', () => {
    const INITIAL_STATE = {
        count: 0,
        data: [],
        history: [],
    }
    expect(reducer(undefined, {})).toEqual(INITIAL_STATE);
});

test('Default reduce', () => {
    const STATE = {
        count: 1,
        data: [],
        history: [1],
    }
    expect(reducer(undefined, {type:types.INCREMENT})).toEqual(STATE);

    const STATE2 = {
        count: 2,
        data: [],
        history: [2,1],
    }
    expect(reducer(STATE, {type:types.INCREMENT})).toEqual(STATE2);

});

test('Default saga', () => {


});

