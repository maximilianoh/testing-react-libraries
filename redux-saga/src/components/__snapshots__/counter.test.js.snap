// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Snap 1`] = `
<div>
  <h1>
     Counter 
  </h1>
  <h1>
     Size list: 
    0
     
  </h1>
  <h1>
     
    0
     
  </h1>
  <button
    onClick={[Function]}
  >
     Increase 
  </button>
  <button
    onClick={[Function]}
  >
     Decrease 
  </button>
  <button
    onClick={[Function]}
  >
     Reset 
  </button>
  <button
    onClick={[Function]}
  >
     Get 
  </button>
</div>
`;
